[단테]  useGetData 혹으로 분리하기 , 나의 현재 위치 데이터 , 현재 위치와 가까운 화장실 위치 마커로 표현하기

Andante23 opened the pull request:
[MOD] types 폴더에 Location 정보 분리하기
[FEAT] home_page/page.tsx 부분에 현재 사용자에서 위치에서 가까운 화장실의 위치 마커로 나타내는 기능 구현


## Task TODOLIST
<!-- 자신이 한 작업을 간단하게 TODO로 표현해주세요! -->
- [x]  types 폴더에  Location 정보를 분리하였습니다.
- [x]  geolocation으로부터 사용자의 위치와 가까운 화장실의 위치 마커로 나타내는 기능을 구현하였습니다.  

## ✨ 개발 내용

(1)  커스텀 혹으로 분리하면  이 로직을 다른 곳에서도 사용이 가능한 재사용이 용이하며 ,해당 로직에 대한 세부 사항을 알 필요가 없다는 것을 통해  컴포넌트의  추상화 및  역활이 명확해집니다. 그리하여 사용하였습니다.

``` ts
//wheres-my-toilet / src / hooks / home_page/useGetData.ts 
export default function useGetData() {
  const [locationInfoData, setLocationInfoData] = useState<Location[] | null>([]);

  async function getData() {
    const { data, error } = await supabase.from('toilet_location').select(' * ');

    setLocationInfoData(data);
    return { data, error };
  }

  useEffect(() => {
    getData();
  }, []);

  return { locationInfoData };
}
```

(2)  지오로케이션 api 를 통해 받아온  나의 현재 위치 데이터가 저장된  zustand 에서 데이터를 가져와서   지도에 표현을 해주었습니다. 

```ts
// zustand 파일
import { create } from 'zustand';

type Store = {
  userLocation: { lat: number; lng: number };
  setLocation: (val: { lat: number; lng: number }) => void;
};
export const useUserLocationStore = create<Store>()((set) => ({
  userLocation: { lat: 37.5560919, lng: 127.146551 }, //임시 기본값
  setLocation: (val) => set({ userLocation: { lat: val.lat, lng: val.lng } }),
}));

//  zustand에 저장되어 있는  사용자 위치데이터 가져오기
  const { userLocation } = useUserLocationStore();
```


## 스크린샷

- 지오로케이션 api를 통해 받아온 나의 현재 위치 데이터가 저장된 zustand 에서 
데이터를 가져와서 지도에 표현한 것을 말합니다. 그리고  현재 나의 위치와  근접한
화장실도 나타내었습니다.
![image](https://github.com/wheres-my-toilet/wheres-my-toilet/assets/98266983/1c0be257-5f83-435e-a3ee-4f6348f9d3ba)

Andante23 committed:
[MOD] useGetData 혹으로 분리하기
[MOD] types 폴더에 Location 정보 분리하기
[FEAT] home_page/page.tsx 부분에 현재 사용자에서 위치에서 가까운 화장실의 위치 마커로 나타내는 기능 구현

Andante23 committed:
Merge branch 'dev' of https://github.com/wheres-my-toilet/wheres-my-toilet into feature/home_page

Andante23 committed:
[MOD] : 불필요한 의존성 추가로 렌더링이 여러번 발생하여 분리

ketchup0211 commented on src/app/home_page/page.tsx:
```diff
@@ -3,45 +3,69 @@
 
-    setLocationInfoData(data);
-    return { data, error };
+  function getDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
```
이 함수는 src 아래에 util 폴더를 만들어 따로 분리하면 더 좋을 것 같아요!

ketchup0211 replied:
입력 값 받을 때 4개를 받지 말고, 객체 형태로 받아 1개의 argument만 있도록 해주세요!

Andante23 replied:
util 에 제가 맡고 있는  페이지이름의 폴더하위에 다음과 같이  생성하였습니다.
![image](https://github.com/wheres-my-toilet/wheres-my-toilet/assets/98266983/e3ddf643-1c71-49e9-a21d-552e3ce4ed26)

Andante23 replied:
```
//  수정된 getDistance 유틸 함수입니다.
import { Distance } from '@/types/home_page/types';

export default function getDistance({ lat1, lon1, lat2, lon2 }: Distance): number {
  const R = 6371; // 지구의 반지름 (단위: km)
  const dLat = (lat2 - lat1) * (Math.PI / 180);
  const dLon = (lon2 - lon1) * (Math.PI / 180);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * (Math.PI / 180)) * Math.cos(lat2 * (Math.PI / 180)) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const distance = R * c; // 두 지점 사이의 거리 (단위: km)
  return distance;
}
```

ketchup0211 commented on src/app/home_page/page.tsx:
```diff
@@ -3,45 +3,69 @@
-    (data) => data.toilet_address?.trim().includes(selectSee) && data.toilet_address?.trim().includes(selectGunGue),
-  );
+    data?.forEach((location) => {
+      const distance = getDistance(
```
여기도 마찬가지로 객체로 전달해주세요!

Andante23 replied:
수정된  findNearestLocation.ts 유틸함수입니다.
```
import { Location, UserLocation } from '@/types/home_page/types';
import getDistance from './getDistance';

export default function findNearestLocation({
  userLocation,
  data,
}: {
  userLocation: UserLocation;
  data: Location[] | null;
}): Location | null {
  let minDistance = Infinity;
  let nearestLocation = null;

  data?.forEach((location) => {
    const distance = getDistance({
      lat1: userLocation.lat,
      lon1: userLocation.lng,
      lat2: location.toilet_latitude,
      lon2: location.toilet_longitude,
    });

    if (distance < minDistance) {
      minDistance = distance;
      nearestLocation = location;
    }
  });

  return nearestLocation;
}

```

ketchup0211 commented:
수고 많으셨습니다! 단테님 최고💪🏻

jigico commented:
기능 구현하느라 시간도 없으실텐데 헤더 컴포넌트 렌더링 이슈까지!! 
이렇게 하나 더 배워갑니다~~!! 👍👍

Nobuki-Kato commented on src/app/home_page/page.tsx:
```diff
@@ -3,45 +3,69 @@
+  lng: number;
+  // 필요한 경우 더 많은 속성 추가 가능
+}
 export default function HomePage() {
```
컨밴션이 지켜지지 않은 것 같아요..! rafce로 작성되지 않은 것 같아요

Andante23 replied:
일단은  저의 homePage 을 rafce 로 바꾸었습니다.!!!! 알려주셔서 감사합니다.
![image](https://github.com/wheres-my-toilet/wheres-my-toilet/assets/98266983/bb580a78-9bf5-4c16-9511-28c3b4f55178)

Nobuki-Kato reviewed (commented):
제일 어렵다고 느껴지는 기능이였는데 해결해 나가시는게 너무 대단해요

Andante23 committed:
[MOD] page.tsx 파일에 있는 코드들 역활 분리
[ADD] 각각 역활에 따른 폴더에  파일 추가

Andante23 reviewed (commented):
Finish your review 변경

Andante23 committed:
[MOD]: util 폴더의  home_page 폴더내에 있는 파일들  수정
[FEAT]: HomeSelectForm 페이지 시 조건에 따라 군/구 선택하게 하는 기능 추가
[FEAT]: 군/구 선택시 선택된 좌표로 지도 확대하는 기능 추가

Andante23 committed:
[ ADD ] 시 군 구 좌표받아오는 함수 파일 service 폴더 > home_page
폴더에 추가

Andante23 committed:
[MOD] useSelectForm import 해서 받아온 함수명 수정

Andante23 committed:
[FIX] HomePageMap 마커 이미지 군/구내 화장실 마커 이미지 제거

Andante23 reviewed (commented):
ttttt

Andante23 commented on src/app/home_page/page.tsx:
```diff
@@ -1,97 +1,58 @@
-import { useEffect, useState } from 'react';
-import { Map, MapMarker } from 'react-kakao-maps-sdk';
-
-interface Location {
```
types/페이지 이름 이렇게 폴더 구조를 만들고  Location을 분리하였습니다.

ketchup0211 reviewed (approved):
클린-코드 너무 좋습니당👍🏻👍🏻 수고하셨어요!
